library/std/src/net/addr.rs

```diff
@@ -290,7 +290,23 @@ impl SocketAddrV4 {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn ip(&self) -> &Ipv4Addr {
-        unsafe { &*(&self.inner.sin_addr as *const c::in_addr as *const Ipv4Addr) }
+        // FIXME: what about other OSes not POSIX and Win32 ?
+
+        // SAFETY: Casting `&in_addr` to `&Ipv4Addr` is safe because:
+        //
+        //   * `Ipv4Addr` is `struct { _: [u8; 4]; }` with size = 4; alignment = 1.
+        //
+        //   * `in_addr` is `struct { _: u32; .. }` in POSIX-compliant OSes [1] and Win32 API [2][3]
+        //      with size and alignment at least 4.
+        //
+        //   * Both:
+        //      * have`#[repr(C)]`.
+        //      * same size (4) and alignment (`align_of in_addr> == 4 * align_of Ipv4Addr`).
```

Comment:
Alignment isn't the same any more, but that's okay -- the key is that `Ipv4Addr` has less restrictive alignment than `self`.